shared/rakefiles/secrets.rb

```diff
@@ -207,6 +227,65 @@ def self.fetch_secrets(encryption_key)
 
     return decrypted_secrets
   end
+
+  # This method creates new primary version for target encryption_key, and
+  # then disables all its previous versions.
+  def self.rotate_key(encryption_key)
+    puts "[secret-mgmt] Creating new primary version for key '#{encryption_key}'..."
+    new_version = %x{
+      gcloud kms keys versions create \
+      --location #{Secrets::KMS_LOCATION} \
+      --keyring #{Secrets::KMS_KEYRING} \
+      --key #{encryption_key} \
+      --primary --format json
+    }
+
+    begin
+      new_version = JSON.parse(new_version)
+      new_version_id = new_version["name"].match(/\/([0-9]+)$/)[1]
+    rescue
+      puts "ERROR: Unable to create new version for key '#{encryption_key}', terminating!"
+      raise
+    end
+
+    puts "[secret-mgmt] Retrieving versions for key '#{encryption_key}'..."
+    key_versions = %x{
+      gcloud kms keys versions list \
+      --location #{Secrets::KMS_LOCATION} \
+      --keyring #{Secrets::KMS_KEYRING} \
+      --key #{encryption_key} \
+      --format json
+    }
+
+    begin
+      key_versions = JSON.parse(key_versions)
+    rescue
+      puts "ERROR: Unable to retrieve versions for key '#{encryption_key}', terminating!"
+      raise
+    end
+
+    key_versions.each do |version|
+      version_id = version["name"].match(/\/([0-9]+)$/)[1]
+      next if version["state"] != "ENABLED" or version_id == new_version_id
+
+      puts "[secret-mgmt] Disabling version #{version_id} for key '#{encryption_key}'..."
+      version_disabled = %x{
+        gcloud kms keys versions disable #{version_id} \
+        --location #{Secrets::KMS_LOCATION} \
+        --keyring #{Secrets::KMS_KEYRING} \
+        --key #{encryption_key} \
+        --format json
+      }
+
+      begin
+        version_disabled = JSON.parse(version_disabled)
+        raise unless version_disabled['state'] == "DISABLED"
+      rescue
+        puts "ERROR: Unable to disable version #{version_id} for key '#{encryption_key}', terminating!"
```

Comment:
If it's safe to do so, it might be helpful to emit the JSON blob as part of this failure to assist debugging.

Reply:
Done.